documentation/adr/0004-trip-modifications-mqtt-topic-structure.md

```diff
@@ -0,0 +1,134 @@
+## Status
+
+Draft
+
+## Context
+
+[With the creation of #ADR0003](./0003-use-mqtt-to-publish-trip-modifications.md),
+we are trying to decide on the structure for _how_ we send and persist these 
+[Trip Modifications](https://gtfs.org/realtime/reference/#message-tripmodifications) between Skate and Transit Data.
+
+We know that we'll need these messages to be available if the Transit Data
+service is restarted.
+
+## Decision
+
+We'll use a
+[topic hierarchy](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/)
+which allows us to manage individual GTFS-RT messages and update them, and does
+not require that the Transit Data service has some way to persist state between
+restarts.
+
+### Topic Hierarchy
+We'll nest all of our Trip Modifications related data under the
+[MQTT "topic level"](https://www.hivemq.com/blog/mqtt-essentials-part-5-mqtt-topics-best-practices/)
+`trip_modifications`, which will be under our "topic prefix"
+configured in the DevOps repo for each environment.
+
+For Example: (spaces used for alignment, and are not part of the topic)
+```
+skate/prod     /trip_modifications/
+skate/dev      /trip_modifications/
+skate/dev-blue /trip_modifications/
+skate/dev-green/trip_modifications/
+```
+
+
+For each new and unique trip modification, we'll generate a unique ID
+(likely `UUID` using `Ecto.UUID.generate()`)
+and use that as our Trip Modification identifier, named `id`.
+
+> [!NOTE]  
+> The GTFS-RT Trip Modifications spec does not currently have a ID for each modification,
+> so this `ID` _will not_ appear with the Trip Modification Message.
+
+The topic hierarchy is then formed using this ID as the topic level
+```
+<prefix>/trip_modifications/#{id}/
+```
+
+Then, to create or update an associated `trip_modification`, we'll use the topic
+level `trip_modification`
+Creating the topic
+```
+<prefix>/trip_modifications/#{id}/trip_modification
+```
+
+Any associated [`Shape`](https://gtfs.org/realtime/reference/#message-shape)
+Messages will be published to the topic level `shape` in the hierarchy.
+```
+<prefix>/trip_modifications/#{id}/shape
+```
+
+On `Shape` Messages, the
+[`Shape.shape_id`](https://gtfs.org/realtime/reference/#message-shape)
+field will also have a unique ID generated by Skate, which will be
+referenced by the corresponding Trip Modification's
+[`SelectedTrips.shape_id`](https://gtfs.org/realtime/reference/#message-selectedtrips)
+field.
+
+#### Active Trip Modifications State
+Messages pushed to these topics will have the
+[`retain` property](https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/)
+set so that new connections get active Trip Modifications and Shapes published
+to them on connect.
+
+Deactivating a Trip Modification will be done by
+["deleting"](https://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages/#heading-how-to-delete-retained-messages-in-mqtt)
+the retained messages on each topic within the `<prefix>/trip_modifications/#{id}/`
```

Comment:
question: does this send a message to current subscribers telling them that the message has been deleted?

Reply:
The mechanism for deleting a retained message is to publish a 0-byte message to the topic, which overwrites the previous retained message. If a subscriber is currently online when that happens, I'd be pretty surprised if they didn't get alerted to the "new 0-byte message".

I'm less certain of what happens if no subscribers are online when we publish the 0-byte message. My guess is that the topic would still be there, but with no retained message (which Transit Data could interpret as "any previously-existing detour with this ID has been deactivated"). But it's possible that the topic itself gets deleted at that point, which means that if Transit Data is having an outage at that point, they wouldn't get that update.

If we decide to go a different route, and signal detour deactivation in some way that leaves a retained message on its MQTT topic, are we worried about cluttering up our MQTT space? Either cognitive load associated with a whole bunch of topics lying around representing ancient detours, or actual space constraints?

And/or is there an easy way to tell MQTT itself to delete topics that haven't had any activity for, say, a week, even if they have a retained message?

Reply:
> question: does this send a message to current subscribers telling them that the message has been deleted?

Great question, this is actually something I wanted to write a test for, thanks for bringing this up! I'll test it in #2679 



> If a subscriber is currently online when that happens, I'd be pretty surprised if they didn't get alerted to the "new 0-byte message".

This is my assumption so far, but will test it 👍🏻.

Reply:
> > If a subscriber is currently online when that happens, I'd be pretty surprised if they didn't get alerted to the "new 0-byte message".
> 
> This is my assumption so far, but will test it 👍🏻.

TL;DR: When a "zero-byte payload" is published, subscribers get a zero byte message.

I tested locally, with Skate code and running `mosquitto_sub` locally.
The MQTT Spec says "publish a retained message with a zero-byte payload" to delete a retained message on a topic. 

To publish a retained "zero-byte payload" with the `EmqttFailover` library, you must send a message that fits the following pattern:
``` elixir
%EmqttFailover.Message{payload: "", retain?: true} = sent_message
```

Subscribers will receive a similar message when this is published, but the retain flag will not be true from what I can tell
``` elixir
%EmqttFailover.Message{payload: "", retain?: true} = received_message
```

When new subscriptions are created though, a message is __not__ sent containing this zero byte payload. So this means the topic is effectively "deleted" for new subscribers, as they'll never know it existed until a new message is published to that topic.

Reply:
> And/or is there an easy way to tell MQTT itself to delete topics that haven't had any activity for, say, a week, even if they have a retained message?

[As of MQTT 5, the spec says that messages can have a Message Expiry Interval](https://www.hivemq.com/blog/mqtt5-essentials-part4-session-and-message-expiry/#heading-message-expiry-interval-in-mqtt-5). This doesn't _seem_ to be exposed in `EmqttFailover`, I don't know if it's exposed in the underlying library either.

Reply:
ActiveMQ 5.17 only supports MQTT 3.

Reply:
I don't think we need Message Expiry anyway, but good to note thank you!